content/blog/2019-12-06-alive/index.md

```diff
@@ -0,0 +1,181 @@
+
+[jr1]: https://blog.regehr.org/archives/1170
+[jr2]: https://blog.regehr.org/archives/1467
+[jr3]: https://blog.regehr.org/archives/1496
+[PR20186]: https://bugs.llvm.org/show_bug.cgi?id=20186
+
+## Evaluating Alive's impact
+At the time of publication in 2015, Alive's authors (manually) ported 334 peephole optimizations.
+Optimizations varied in verification time from a few seconds to several hours.
+From these 334 optimizations, Alive found 8 bugs.
+
+In addition, the authors build a version of LLVM with the default instruction combiner replaced by Alive-generated C++ for their 334 optimizations.
+They found that despite not covering all of the previous optimizations, LLVM+Alive maintained within 10% of the performance of LLVM on SPEC 2000 and 2006 benchmarks.
+Much more interestingly, however, the authors show how little coverage these optimizations received in the existing tests and benchmarks.
+An instrumented LLVM-Alive run on LLVM's nightly test suite and both SPEC benchmarks found that only 159 of the 334 optimizations were triggered:
+
+<img src="llvm-alive.png" width="500" >
+
+That is, nearly half of the peephole optimizations ported to Alive were untested via the existing manual test and benchmark flow!
+
+In addition to their hard performance numbers, Alive's authors reached out to LLVM developers to incorporate Alive into work-in-progress patches.
+The authors report they found "dozens" of proposed incorrect optimization implementations, which they were able to provide counter-examples to prevent with the help of Alive.
+
+## Key take-aways
+Alive leaves us with several key nuggets of wisdom:
+
+#### *DSL + SMT = profit*
+Alive demonstrates that finding a domain-specific language for your goals, in this case concise peephole optimizations, can be especially fruitful for verification.
+The authors argue that DSLs help compiler engineers reason about code.
+Beyond that, Alive shows that a DSL makes translation of semantics to a formal logic like SMT more tractable than trying to wrangle with the full semantics of languages like C or LLVM IR directly.
+Later work on Alive (["Alive2"][a2]) has also introduced tools to help translate LLVM IR to Alive's DSL in an automated fashion.
+
+[a2]: https://github.com/AliveToolkit/alive2
+
+#### *Usability matters in formal methods*
+
+Alive is a formal system, but it is also a deeply practical one.
+It recognized that there is impact to be had from building verification systems closer to where working programmers spend their day-today-hacking, in part by targeting a massive existing code base in a piecewise, workable way.
+In addition, Alive's DSL and counter-examples were designed with an interface meant to be familiar to LLVM engineers, which undoubtedly paid off in the adoption of this work.
+Finally, the authors of Alive engaged closely with the LLVM community, from frequenting the RFC discussion channels to publishing high-level blog posts on their contributions.
```

Comment:
One more depressing lesson, however, might be that—even despite this deep engagement and adherence to usability—LLVM still has not wholesale replaced most of InstCombine with Alive-generated code. Real-world tech transfer is super hard!